### PR TITLE
Add docs for preventing volume mode conversion

### PR DESCRIPTION
**Pull request by RaunakShah**
Documentation changes for feature to prevent unauthorised volume mode conversion

KEP - https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3141-prevent-volume-mode-conversion

### CODE REVIEW

**Comment by RaunakShah** [comment]
cc @xing-yang

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [inline content/en/docs/concepts/storage/volume-snapshots.md:152]: It's possible that a user may be reading this doc but is using an older version of CRDs and controllers.  So we need to add notes right next to this field to clarify how it can be used.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [inline content/en/docs/concepts/storage/volume-snapshots.md:123]: It's possible that a user may be reading this doc but is using an older version of CRDs and controllers. So we need to add notes right next to this field to clarify how it can be used.

Comment (RaunakShah) [reply content/en/docs/concepts/storage/volume-snapshots.md]: Done!

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:123]: > This field is only present for release-6.0 on VolumeSnapshotContent API.

If I use a later release, the text implies I can only use `Filesystem`. Am I right that it's SemVer and that 6.x and any later major version support the `Block` mode?
(we should make the text unambiguous; that also helps localization teams)

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:157]:
```suggestion
If you want to allow users to create a `PersistentVolumeClaim` from an existing `VolumeSnapshot`,
the annotation `snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"` needs to be added
to the `VolumeSnapshotContent` that corresponds to the `VolumeSnapshot`. For example:
```

Notice that I made the annotation value be a string not a bool.

**Review by sftim — DISMISSED** [review]
This PR also needs to register the `snapshot.storage.kubernetes.io/allowVolumeModeChange` annotation, by updating https://kubernetes.io/docs/reference/labels-annotations-taints/

Please add that detail. Also, I spotted an incorrect annotation value.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-snapshots.md:123]: SourceVolumeMode is the mode of the volume whose snapshot is taken. It can be either “Filesystem” or “Block”. If not specified, it indicates the source volume's mode is unknown. This field is available in Volume Snapshot client v6.0.0 and higher.

Comment (sftim) [reply content/en/docs/concepts/storage/volume-snapshots.md:123]: @xing-yang what change should @RaunakShah make so that the situation is clear to readers?

Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-snapshots.md:123]: I'm suggesting to add these https://github.com/kubernetes/website/pull/32673#discussion_r846635408 as comments next to the sourceVolumeMode.  Any other suggestions?

**Comment by nate-double-u** [comment]
Hello, this is a friendly reminder from the 1.24 docs release team that the _Docs' complete deadline_ — All PRs reviewed and ready to merge — is EOD **Tuesday, April 12th, 2022**. Please finish up any remaining technical reviews, and reach out if you need any help.

**Review by sftim — CHANGES_REQUESTED** [review]
This PR needs to register the snapshot.storage.kubernetes.io/allowVolumeModeChange annotation, by updating https://kubernetes.io/docs/reference/labels-annotations-taints/

**Comment by sftim** [comment]
The PR mentions an unregistered annotation. We will note the tech LGTM for further reviews.
/lgtm cancel

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:152]:
```suggestion
`sourceVolumeMode` is the mode of the volume whose snapshot is taken. The value of the `sourceVolumeMode` field can be either `Filesystem` or `Block`. If you do not specify a source volume mode, Kubernetes treats the snapshot as if the source volume's mode is unknown. Support for this field is available in VolumeSnapshot client v6.0.0 and higher.
```

If someone asked me for the URL of the VolumeSnapshot client, I would not have a clue what that URL is or how to work it out. We should make it clear what we mean by the phrase “VolumeSnapshot client v6.0.0”.

Comment (RaunakShah) [reply content/en/docs/concepts/storage/volume-snapshots.md]: Thanks for the suggestion! I've used most of it but modified `If you do not specify a source volume mode` to `If the source volume mode is not specified`. It is not neccesarily on the user to populate this field.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [inline content/en/docs/concepts/storage/volume-snapshots.md:153]: Add a link to https://github.com/kubernetes-csi/external-snapshotter/releases

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (tengqm) [inline content/en/docs/reference/labels-annotations-taints/_index.md:505]:
```suggestion
See [Converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/) for more information.
```

Comment (tengqm) [reply content/en/docs/reference/labels-annotations-taints/_index.md:505]: The link target doesn't have the `.md` suffix.

Comment (RaunakShah) [reply content/en/docs/reference/labels-annotations-taints/_index.md]: fixed!

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:155]: nit:
```suggestion
## Converting the volume mode of a snapshot {#convert-volume-mode}
```

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (sftim) [inline content/en/docs/reference/labels-annotations-taints/_index.md:498]: I'd have this as:
```suggestion
Example: `snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"`
```

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (sftim) [inline content/en/docs/reference/labels-annotations-taints/_index.md:503]: I'd write:
```suggestion
This determines whether a user can modify the mode of the source volume when a 
{{< glossary_tooltip text="PersistentVolumeClaim" term_id="persistent-volume-claim" >}} is being
created from a VolumeSnapshot.
```

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:153]: Do you need to be using Kubernetes v1.24 to set this field?

**Review by sftim — COMMENTED** [review]
Some more feedback.

Also:
- should this change actually target the _dev-1.24_ branch?
- is the `sourceVolumeMode` field added unconditionally, or it it behind a feature gate?

**Comment by sftim** [comment]
/milestone 1.24

Feel free to unset this if I've misunderstood.

**Comment by RaunakShah** [comment]
> should this change actually target the dev-1.24 branch?

Can you clarify what the criteria is to target main vs a dev branch?

> is the sourceVolumeMode field added unconditionally, or it it behind a feature gate?

The field is present is part of the CRD unconditionally. There's an out-of-tree feature flag that needs to be enabled; I'm not sure if that's generally documented here or in that specific component.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (RaunakShah) [reply content/en/docs/concepts/storage/volume-snapshots.md:153]: `VolumeSnapshot`s 6.0 release will only be supported with a minimum k8s version of 1.24. @xing-yang can correct me if I'm wrong here.

**Comment by sftim** [comment]
Ah, I'd missed that this is a custom resource. What version of that CustomResourceDefinition do people need to deploy?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [reply content/en/docs/concepts/storage/volume-snapshots.md:153]: Do we mean:
> you need to be using `external-snapshotter` release 6.0 or later (which then requires Kubernetes v1.24)

?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:157]: > This feature is only present for `VolumeSnapshotContents` created with client version `v6.0.0` onwards.

I'm guessing that a client can examine the `volumesnapshotcontents` API to see if it includes a `sourceVolumeMode` field; is that the only way for a client to tell whether the feature is supported?

**Comment by RaunakShah** [comment]
Between 
> you need to be using external-snapshotter release 6.0 or later (which then requires Kubernetes v1.24)

and

> What version of that CustomResourceDefinition do people need to deploy?

The CRD and external-snapshotter version need to be v6.0.0 onwards, but the feature flag only needs to be enabled on external-snapshotter. 

There's also a blog PR for this feature (https://github.com/kubernetes/website/pull/32419)  where I thought I would clarify the exact way to enable the feature. Do you think its better to be mentioned here, instead of just pointing to the release page?

**Comment by xing-yang** [comment]
> should this change actually target the dev-1.24 branch?

@sftim Good catch.  I missed that.  @RaunakShah Please submit another PR with this content to branch dev-1.24.  Since this is targeting 1.24, the doc PR should be submitted to that branch instead of main.

> is the sourceVolumeMode field added unconditionally, or it it behind a feature gate?

SourceVolumeMode field is added unconditionally.  VolumeSnapshot APIs are CRDs.  We don't have a way to control it at API server level.  The feature gate will be on the snapshot controller.

**Comment by sftim** [comment]
@RaunakShah could you rebase and squash these commits (against _dev-1.24_)?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-snapshots.md:153]: VolumeSnapshot APIs are CRDs. So this field is not really in k/k code base.  
The only thing that is in k/k code base for the snapshot feature is the DataSource field in PVC which moved to GA in 1.20 release.
But we do update K8s dependencies in https://github.com/kubernetes-csi/external-snapshotter code and release it shortly after every K8s release. VolumeSnapshot client will be released shortly after K8s 1.24 release.

**Review by sftim — COMMENTED** [review]
I'm concerned that I have been documenting Kubernetes for a while and yet I don't feel like I understand the new documentation here.

If it's confusing me, I'm concerned that it will confuse other readers.

**Comment by xing-yang** [comment]
> I'm concerned that I have been documenting Kubernetes for a while and yet I don't feel like I understand the new documentation here.
> If it's confusing me, I'm concerned that it will confuse other readers.

@sftim I'm looking at K8s docs on volume snapshots: https://kubernetes.io/docs/concepts/storage/volume-snapshots/#introduction).  It doesn't talk about how to deploy volume snapshot CRDs and controller.  It has a link to CSI driver docs: https://kubernetes-csi.github.io/docs/
Should we just add a brief introduction of this feature and refer to CSI driver docs for details?

**Comment by nate-double-u** [comment]
Deploy previews:
* https://deploy-preview-32673--kubernetes-io-main-staging.netlify.app/docs/concepts/storage/volume-snapshots/#volume-snapshot-contents
* https://deploy-preview-32673--kubernetes-io-main-staging.netlify.app/docs/reference/labels-annotations-taints/#snapshot-storage-kubernetes-io-allowvolumemodechange

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (xing-yang) [inline content/en/docs/reference/labels-annotations-taints/_index.md:521]: @sftim Does it make sense to only keep this brief section for labels-annotations-taints and refer to https://kubernetes-csi.github.io/docs/ for details?

**Comment by nate-double-u** [comment]
Hello, this is a friendly reminder that the docs ready deadline has passed. Since we moved the release date to May 3, I'm OK with the team taking some time updating this PR to make sure that it is ready before merging; however, I would like to see this merged in by Thursday, April 21.

Is there anything the docs release team or I can do to help unstick this PR?

**Comment by xing-yang** [comment]
@nate-double-u Does doc team want this PR to be updated with less details (leaving details in CSI docs https://kubernetes-csi.github.io/docs/) or are you fine with the current PR?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (divya-mohan0209) [reply content/en/docs/concepts/storage/volume-snapshots.md:153]: What we're trying to understand here is if there is a dependency for the version of `VolumeSnapshot`client on the Kubernetes release . You've mentioned that only the Datasource field (which moved to GA) is there in the k/k codebase. Are the other K8s dependencies that you mentioned related to other components within a release even though they're out of tree?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (divya-mohan0209) [reply content/en/docs/concepts/storage/volume-snapshots.md:157]: Only speaking for myself here, but I think this would be good to elaborate @RaunakShah .

**Review by divya-mohan0209 — COMMENTED** [review]
A few suggestions.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-snapshots.md:153]: VolumeSnapshot client uses a few K8s components.  See here:
https://github.com/kubernetes-csi/external-snapshotter/blob/v6.0.0-rc4/client/go.mod#L6-L9
Currently it uses 1.23 components. After K8s 1.24 is released, the dependencies will be updated to 1.24 and we'll cut the official v6.0.0 client release.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-snapshots.md:157]: How about this: `This feature is in go module github.com/kubernetes-csi/external-snapshotter/client/v6 and later.`
I think a user can check if the new field `sourceVolumeMode` exists in a VolumeSnapshotContent spec or not by checking the manifest.

Comment (sftim) [reply content/en/docs/concepts/storage/volume-snapshots.md:157]: In these docs, we almost never talk about Go modules. Do end users need to look at that level of detail to make use of snapshots?

Imagine someone is doing development with a bit of operations, and their thoughts run like this:
- I have access to a Kubernetes cluster cluster
- There is a team who look after the platform, including that cluster
- I want to know whether the cluster lets me use `sourceVolumeMode`
- I know I can write a test to see if `sourceVolumeMode` works in my cluster and then use the result of that test
- I know I can ask the platform team to do the same test and let me know the outcome
- if `sourceVolumeMode` is available, I want to use it. But only in the development environment.

Comment (sftim) [reply content/en/docs/concepts/storage/volume-snapshots.md:157]: Can that developer write code that uses the Kubernetes API to work out whether `sourceVolumeMode` is available (eg by querying the cluster's OpenAPI data)?

If they can, let's give them a clear hint about that. Much better than to tell them to read the source code for a snapshotting component.
If the easiest way to work out what's available is to look at the Pod template for an add-on (`external-snapshotter`?) then we should tell them what to look for in that Pod template.

**Comment by sftim** [comment]
Please see feedback / query in https://github.com/kubernetes/website/pull/32673#discussion_r852980969

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-snapshots.md:157]: They can check if `sourceVolumeMode` is available in the manifest using kubectl: `kubectl get crd volumesnapshotcontent -o yaml`.
@RaunakShah Can you update the PR with this?

**Comment by nate-double-u** [comment]
> @nate-double-u Does doc team want this PR to be updated with less details (leaving details in CSI docs https://kubernetes-csi.github.io/docs/) or are you fine with the current PR?

Hi @xing-yang, as the release docs lead I'm looking for two `/lgtm`s, one technical (which yours is), and one from the sig-docs team. I'd prefer not to `/approve` before getting both.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [inline content/en/docs/concepts/storage/volume-snapshots.md:157]: Can you remove v6.0.0 as it is confusing for users?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (xing-yang) [inline content/en/docs/concepts/storage/volume-snapshots.md:161]: Remove 6.0.0 here as well.

**Comment by RaunakShah** [comment]
> Please see feedback / query in https://github.com/kubernetes/website/pull/32673#discussion_r852980969

@sftim I've removed all references to `v6.0.0` in the latest commit and used the suggestion by @xing-yang in this comment - https://github.com/kubernetes/website/pull/32673#discussion_r853046270

Can you review this again?

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (divya-mohan0209) [reply content/en/docs/reference/labels-annotations-taints/_index.md:518]: @RaunakShah : Please can you have a look at this suggestion? It doesn't seem to have been addressed and we think it would add value.

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (divya-mohan0209) [reply content/en/docs/concepts/storage/volume-snapshots.md:123]: @RaunakShah : This conversation was marked as resolved and I did not see @xing-yang 's comments actually addressed. Could you clarify if this was agreed upon to not be added eventually, please?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (divya-mohan0209) [reply content/en/docs/concepts/storage/volume-snapshots.md:152]: I don't see this addressed either, ptal.

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (divya-mohan0209) [reply content/en/docs/reference/labels-annotations-taints/_index.md:513]: @RaunakShah : PTAL since this hasn't been addressed and we believe it improves readability of the doc.

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (divya-mohan0209) [reply content/en/docs/reference/labels-annotations-taints/_index.md:521]: @xing-yang : From a personal standpoint, I'd err on the side of too much information rather than too less. 
In the spirit of what I just said, what we could potentially rewrite it as is suggested above.

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (divya-mohan0209) [inline content/en/docs/reference/labels-annotations-taints/_index.md:520]:
```suggestion
Refer to the documentation on [Converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/#convert-volume-mode) and the [Kubernetes CSI Developer Documentation](https://kubernetes-csi.github.io/docs/) for more information.
```
```suggestion
See [Converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/#convert-volume-mode) for more information.
```

Comment (divya-mohan0209) [inline content/en/docs/reference/labels-annotations-taints/_index.md:520]:
```suggestion
Refer to [Converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/#convert-volume-mode) and the [Kubernetes CSI Developer Documentation](https://kubernetes-csi.github.io/docs/) for more information.
```

**Review by divya-mohan0209 — CHANGES_REQUESTED** [review]
Added some feedback and unresolved some comments since I did not find them addressed. PTAL and let us know, if any queries.

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (RaunakShah) [reply content/en/docs/reference/labels-annotations-taints/_index.md]: Done

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (RaunakShah) [reply content/en/docs/reference/labels-annotations-taints/_index.md]: Done

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (RaunakShah) [reply content/en/docs/reference/labels-annotations-taints/_index.md]: Done

File: content/en/docs/reference/labels-annotations-taints/_index.md
Comment (sftim) [reply content/en/docs/reference/labels-annotations-taints/_index.md:521]: Which specific page on https://kubernetes-csi.github.io/docs/ would we link to?

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:168]: - this isn't YAML
- [don't include the command prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt)

File: content/en/docs/concepts/storage/volume-snapshots.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-snapshots.md:169]: After running `kubectl get crd volumesnapshotcontent`, does a success exit indicate support for the  sourceVolumeMode` field? If not, we should document how to actually check.

**Review by sftim — DISMISSED** [review]

**Comment by nate-double-u** [comment]
Hi Everyone!

I appreciate the effort going into this, but this PR has a little less than a week to be resolved and merged into `dev-1.24`; everything needs to be complete and merged in by next Thursday (**April 28, 2022**) as we will be freezing the k/website repo on the following Monday.

/cc @kubernetes/sig-docs-leads @JamesLaverack @xing-yang

**Comment by sftim** [comment]
As this is an alpha feature, perhaps we can update the release notes with a comment to make it clear that the docs for it are not finished, ship what we have, and then make sure that we don't graduate the feature to beta until they're improved?

**Comment by divya-mohan0209** [comment]
I'm normally not on the side of holding back a feature because of a doc. However I do feel like this doc needs a lot of work and  possibly, a rewrite. So +1 to the suggestion by @sftim if the feature owner/release lead are okay with this suggestion.

**Comment by nate-double-u** [comment]
+1 to @sftim's suggestion. I'm happy to approve this if either you or @divya-mohan0209 give this an `/lgtm` _and_ @AuraSinis agrees we can put this into the release notes.

@AuraSinis, could you provide next steps on adding an update to the release notes?

**Comment by AuraSinis** [comment]
If no action is required I can just put in a note about docs still in-situ manually, easy-peasy.

**Comment by nate-double-u** [comment]
Thansk @AuraSinis.
@RaunakShah, @xing-yang, do you agree with this path?

**Comment by xing-yang** [comment]
@nate-double-u That's fine with me.

**Comment by RaunakShah** [comment]
@nate-double-u this works for me too

**Comment by nate-double-u** [comment]
Thanks everyone, I appreciate all the effort you've put in!

/lgtm
/approve

**Pull request merged**
